File: packages/shared/src/storage/activity-logger.ts
/**
 * Activity Logger
 *
 * Tracks note editing activity for cross-instance synchronization.
 * Each instance maintains its own log file to avoid multi-writer conflicts.
 */

import type { FileSystemAdapter } from './types';

export class ActivityLogger {
  private activityLogPath: string;
  private profileId = '';

  constructor(
    private fs: FileSystemAdapter,
    private activityDir: string
  ) {
    // IDs will be passed via setIds() after construction
    this.activityLogPath = '';
  }

  /**
   * Set profile and instance IDs and initialize log path
   *
   * Filename format: {profileId}.{instanceId}.log
   * Uses '.' as delimiter since profileId can contain '_' (base64url alphabet).
   *
   * Old formats ({instanceId}.log or {profileId}_{instanceId}.log) are still
   * readable for backward compatibility in activity-sync.ts.
   */
  setIds(profileId: string, instanceId: string): void {
    this.profileId = profileId;
    // instanceId is only used for the filename, not stored separately
    this.activityLogPath = this.fs.joinPath(this.activityDir, `${profileId}.${instanceId}.log`);
  }

  /**
   * @deprecated Use setIds() instead
   */
  setInstanceId(instanceId: string): void {
    // Legacy support - uses instanceId as both profile and instance
    this.setIds(instanceId, instanceId);
  }

  /**
   * Initialize activity logger (ensure directory exists)
   */
  async initialize(): Promise<void> {
    await this.fs.mkdir(this.activityDir);
  }

  /**
   * Record note activity
   *
   * Always appends a new line for each update to ensure other instances
   * see all intermediate sequence numbers during incremental sync.
   *
   * Format: noteId|profileId|sequenceNumber
   * Uses '|' as delimiter for all fields since profileId can contain '_'.
   *
   * Old formats (with '_' delimiter) are still readable in activity-sync.ts.
   *
   * Note: The compact() function prevents unbounded growth by keeping
   * only the last 1000 entries.
   */
  async recordNoteActivity(noteId: string, sequenceNumber: number): Promise<void> {
    const line = `${noteId}|${this.profileId}|${sequenceNumber}`;

    // Always append - don't use replaceLastLine optimization
    // This ensures other instances see ALL intermediate sequences, which is
    // critical when cloud sync services (iCloud, Dropbox) sync files incrementally
    await this.appendLine(line);
  }

  /**
   * Append a line to the activity log
   */
  private async appendLine(line: string): Promise<void> {
    try {
      // Read existing content
      const existingData = await this.fs.readFile(this.activityLogPath);
      const existingText = new TextDecoder().decode(existingData);
      const newText = existingText + line + '\n';
      const newData = new TextEncoder().encode(newText);
      await this.fs.writeFile(this.activityLogPath, newData);
    } catch {
      // File doesn't exist, create it
      const newData = new TextEncoder().encode(line + '\n');
      await this.fs.writeFile(this.activityLogPath, newData);
    }
  }

  /**
   * Compact activity log to keep only the last N entries
   *
   * This prevents unbounded file growth. Safe to call from any instance
   * because each instance only modifies its own log file.
   */
  async compact(maxEntries = 1000): Promise<void> {
    try {
      const data = await this.fs.readFile(this.activityLogPath);
      const content = new TextDecoder().decode(data);
      const lines = content.split('\n').filter((l) => l.length > 0);

      if (lines.length > maxEntries) {
        const kept = lines.slice(-maxEntries);
        const newContent = kept.join('\n') + '\n';
        const newData = new TextEncoder().encode(newContent);
        await this.fs.writeFile(this.activityLogPath, newData);
      }
    } catch (error) {
      // File might not exist yet, that's fine
      console.error('[ActivityLogger] Failed to compact:', error);
    }
  }

  /**
   * Get the path to this instance's activity log
   */
  getLogPath(): string {
    return this.activityLogPath;
  }
}

File: packages/ios/NoteCove/Storage/ActivityLogger.swift
/**
 * Activity Logger
 *
 * Tracks note editing activity for cross-instance synchronization.
 * Each instance maintains its own log file to avoid multi-writer conflicts.
 *
 * Log format: noteId|profileId|sequenceNumber
 */

import Foundation

@MainActor
class ActivityLogger {
    static let shared = ActivityLogger()

    private var profileId: String = ""
    private var instanceId: String = ""
    private var activityDir: URL?

    /// Tracks the last sequence number used for each note
    /// Key: noteId, Value: last sequence number
    private var sequenceNumbers: [String: Int] = [:]

    private init() {}

    /// Configure the activity logger with profile and instance IDs
    /// - Parameters:
    ///   - profileId: The profile ID (SD ID)
    ///   - instanceId: The instance ID for this device
    func configure(profileId: String, instanceId: String) {
        self.profileId = profileId
        self.instanceId = instanceId

        // Get the activity directory from the active SD
        if let activeDir = StorageDirectoryManager.shared.activeDirectory,
           let sdURL = activeDir.url {
            self.activityDir = sdURL.appendingPathComponent(".notecove").appendingPathComponent("activity")

            // Ensure directory exists
            try? FileManager.default.createDirectory(at: activityDir!, withIntermediateDirectories: true)
        }

        print("[ActivityLogger] Configured with profileId=\(profileId), instanceId=\(instanceId)")
    }

    /// Get the next sequence number for a note
    /// - Parameter noteId: The note ID
    /// - Returns: The next sequence number to use
    func getNextSequenceNumber(for noteId: String) -> Int {
        let current = sequenceNumbers[noteId] ?? 0
        let next = current + 1
        sequenceNumbers[noteId] = next
        return next
    }

    /// Record note activity to the log file
    /// - Parameters:
    ///   - noteId: The note ID that was modified
    ///   - sequenceNumber: The sequence number of this update
    func recordNoteActivity(noteId: String, sequenceNumber: Int) {
        guard !profileId.isEmpty, !instanceId.isEmpty else {
            print("[ActivityLogger] Not configured, skipping activity recording")
            return
        }

        guard let activityDir = activityDir else {
            print("[ActivityLogger] Activity directory not set")
            return
        }

        // Filename format: {profileId}.{instanceId}.log
        let filename = "\(profileId).\(instanceId).log"
        let fileURL = activityDir.appendingPathComponent(filename)

        // Line format: noteId|profileId|sequenceNumber
        let line = "\(noteId)|\(profileId)|\(sequenceNumber)\n"

        do {
            if FileManager.default.fileExists(atPath: fileURL.path) {
                // Append to existing file
                let handle = try FileHandle(forWritingTo: fileURL)
                handle.seekToEndOfFile()
                if let data = line.data(using: .utf8) {
                    handle.write(data)
                }
                handle.closeFile()
            } else {
                // Create new file
                try line.write(to: fileURL, atomically: true, encoding: .utf8)
            }

            print("[ActivityLogger] Recorded activity: noteId=\(noteId), seq=\(sequenceNumber)")
        } catch {
            print("[ActivityLogger] Failed to record activity: \(error)")
        }
    }

    /// Get the path to this instance's activity log
    func getLogPath() -> URL? {
        guard let activityDir = activityDir else { return nil }
        let filename = "\(profileId).\(instanceId).log"
        return activityDir.appendingPathComponent(filename)
    }
}
